src/d3a/models/market/market_structures.py

```diff
@@ -34,75 +35,97 @@ class Clearing:
         self.price = price
         self.energy_rate = self.price / self.energy
 
-    def __repr__(self):
-        return ("<Offer('{s.id!s:.6s}', '{s.energy} kWh@{s.price}', '{s.seller} {rate}'>"
-                .format(s=self, rate=self.energy_rate))
-
-    def __str__(self):
-        return ("{{{s.id!s:.6s}}} [origin: {s.seller_origin}] "
-                "[{s.seller}]: {s.energy} kWh @ {s.price} @ {rate}"
-                .format(s=self, rate=self.energy_rate))
+    def update_energy(self, energy: float) -> None:
+        self.energy = energy
+        self.energy_rate = self.price / self.energy
 
-    def to_json_string(self, **kwargs):
-        """Convert the Offer object into its JSON representation.
+    def to_json_string(self, **kwargs) -> str:
+        """Convert the Offer or Bid object into its JSON representation.
 
         Args:
             **kwargs: additional key-value pairs to be added to the JSON representation.
         """
-        offer_dict = deepcopy(asdict(self))
+        obj_dict = deepcopy(self.__dict__)
         if kwargs:
-            offer_dict = {**offer_dict, **kwargs}
+            obj_dict = {**obj_dict, **kwargs}
+
+        obj_dict["type"] = self.__class__.__name__
 
-        offer_dict["type"] = "Offer"
-        offer_dict.pop('energy_rate', None)
+        return json.dumps(obj_dict, default=my_converter)
 
-        return json.dumps(offer_dict, default=my_converter)
+    def _to_csv(self) -> Tuple:
+        rate = round(self.energy_rate, 4)
+        return (rate, self.energy, self.price,
+                self.seller if isinstance(self, Offer) else self.buyer)
 
-    def serializable_dict(self):
+    def serializable_dict(self) -> Dict:
         return {
-            "type": "Offer",
+            "type": self.__class__.__name__,
             "id": self.id,
             "energy": self.energy,
             "energy_rate": self.energy_rate,
-            "original_offer_price": self.original_offer_price,
-            "seller": self.seller,
-            "seller_origin": self.seller_origin,
-            "seller_origin_id": self.seller_origin_id,
-            "seller_id": self.seller_id,
             "time": datetime_to_string_incl_seconds(self.time),
             "attributes": self.attributes,
             "requirements": self.requirements
         }
 
-    def __hash__(self):
+
+class Offer(BaseBidOffer):
+    def __init__(self, id: str, time: datetime, price: float,
+                 energy: float, seller: str, original_offer_price: float = None,
+                 seller_origin: str = None, seller_origin_id: str = None,
+                 seller_id: str = None,
+                 attributes: Dict = None,
+                 requirements: List[Dict] = None):
+        super().__init__(id=id, time=time, price=price, energy=energy,
+                         attributes=attributes, requirements=requirements)
+        self.seller = seller
+        self.original_offer_price = original_offer_price
```

Comment:
As I suggested on slack, this could also be renamed to be 
```suggestion
        self.original_price = original_offer_price
```
and put into `BidsOffer`